[UXUI] Enhance phone field with country-specific validation

andersonjeccel opened the pull request:
| Q                                      | A
| -------------------------------------- | --------------------------
| Bug fix?          | ❌
| New feature/enhancement?      | ✔️
| Deprecations?                          | ❌
| BC breaks?        | ❌
| Automated tests included?              | ✔️ 
| Related user documentation PR URL      | 
| Related developer documentation PR URL | 
| Issue(s) addressed                     | Fixes #1614

## Description

This PR enhances the phone field in Mautic forms by adding country-specific validation capabilities. The feature addresses the long-standing issue where phone fields accepted any input without proper validation, as reported in issue #1614.

This enhancement significantly improves the reliability of phone number data collected through Mautic forms while maintaining the flexibility and ease of use that Mautic users expect.

[Based on this GitHub repository](https://github.com/ariankoochak/regex-patterns-of-all-countries/blob/main/README.md)

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a new form or edit an existing form
3. Add a phone field to the form
4. In the phone field settings, verify the new "Country-specific validation" dropdown is present
5. Select a country from the dropdown (e.g., United States, United Kingdom, Germany)
6. Optionally customize the validation message
7. Save the form
8. Test the form by entering valid and invalid phone numbers for the selected country, **using international format (with country code, for example, +55, +32 etc)**
9. Verify that validation works correctly and appropriate error messages are displayed
10. Test with different countries to ensure validation patterns are country-specific

## Screenshot

![Phone field country validation settings](https://github.com/user-attachments/assets/f762d50f-c14a-4f25-904f-41d471274856)

andersonjeccel committed:
Enhance phone field with country validation

andersonjeccel committed:
Move tag tooltip to wrapper div

andersonjeccel committed:
Use single trans string for tooltip

andersonjeccel committed:
Make country names translatable

andersonjeccel committed:
Apply PHP rules

imaabasiee reviewed (commented):
<img width="1250" height="390" alt="image" src="https://github.com/user-attachments/assets/cb6aa1f0-8bbf-4f97-bdda-7a414bc0c88f" />
Validation does not work correctly. Tested with both a correct and an incorrect Nigerian and UK numbers and got same error message in all scenarios.

I think for the country code you should specify if the phone number should ideally include the country code (e.g +234) to ensure it's validation or auto fills when a Nigerian number is entered without the country code or prompt the user to add it or maybe this is already done in the backend? I just want to know if I enter a Nigerian phone number with or without the country code, the form can successfully go through. N/B: This applies to any specific country selected.

jos0405 commented:
Hi,
Validation didn't work for:

<img width="1051" height="376" alt="image" src="https://github.com/user-attachments/assets/0d9e68b2-2e0a-4ee4-a3c6-61fc54dbe043" />

PLUS:
I think the validation messages should not be directly under each other.
Connfusing to understand which is which.

<img width="935" height="824" alt="image" src="https://github.com/user-attachments/assets/f762d530-0108-4eb8-9c6c-a01031c4e3c9" />

jos0405 reviewed (changes requested):
See above

kuzmany commented on app/bundles/FormBundle/Helper/PhoneCountryValidationHelper.php:
Mautic use library vendor/giggsey/libphonenumber-for-php for handle phone number in PHP, for example here app/bundles/CoreBundle/Helper/PhoneNumberHelper.php. Then we do not need increase codebase with duplicate codes.

Then validation should looks like - example:

```
use libphonenumber\PhoneNumberUtil;

  $phoneUtil = PhoneNumberUtil::getInstance();

  try {
      $number = $phoneUtil->parse('612345678', 'ES');
      $isSpanish = $phoneUtil->isValidNumberForRegion($number, 'ES');

      echo $isSpanish ? 'Valid Spanish number' : 'Not a Spanish number';

  } catch (\libphonenumber\NumberParseException $e) {
      echo 'Invalid number';
  }
```

kuzmany replied:
The library has methods to get all supported countries:

```
  use libphonenumber\PhoneNumberUtil;

  $phoneUtil = PhoneNumberUtil::getInstance();

  // Get all supported region codes (e.g., ['US', 'ES', 'SK', 'GB', 'FR', ...])
  $regions = $phoneUtil->getSupportedRegions();

  // Example: Display as dropdown options
  foreach ($regions as $regionCode) {
      echo $regionCode . "\n"; // US, ES, SK, etc.
  }
```

  The getSupportedRegions() method returns an array of all country codes (like 'ES', 'SK', 'US', 'GB', etc.) that the
  library supports.

andersonjeccel replied:
Definitely useful, thank you for the idea!
Implemented

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Move solution to library

andersonjeccel committed:
Improve UI

andersonjeccel committed:
csfix

RCheesley committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
merge 7.x and cover phone validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
fix phone validation label

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
fix phone field test sonar issue

andersonjeccel committed:
update branch with 7.x

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

kuzmany committed:
Store ISO region code for phone country validation instead of locale-dependent name

The country choice persisted the localized display name (e.g. "United States" /
"États-Unis"), resolved via Symfony Intl Countries::getNames() under the current
locale. A form built under one locale and submitted under another produced a name
that no longer matched, so getCountryCodeFromName() returned null and country
validation was silently skipped — every number was accepted regardless of country.

Store the stable ISO region code (e.g. "US") as the choice value instead; the
dropdown still shows the localized label. getCountryCodeFromName() is removed and
the subscriber consumes the code directly, guarding unknown/legacy values via
Countries::exists() so they skip validation rather than crash or reject everything.

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

kuzmany reviewed (approved):
I identified an issue with using country names in languages other than English. I corrected it, and the implementation now appears to function correctly.

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
fix rector

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
fix rector

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
fix rector

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

andersonjeccel committed:
Merge branch '7.x' into feature/Enhance-phone-field-with-country-validation

shinde-rahul commented on app/bundles/FormBundle/EventListener/FormValidationSubscriber.php:
```diff
@@ -113,19 +115,68 @@ private function fieldEmailValidation(Events\ValidationEvent $event): void
+            return false;
+        }
+
+        $message = $validation['country_validationmsg'] ?? $this->translator->trans('mautic.form.submission.phone.invalid_country', ['%country%' => Countries::getName($countryCode)], 'validators');
```
```suggestion
        $message = !empty($validation['country_validationmsg'])
            ? $validation['country_validationmsg']
            : $this->translator->trans(
                'mautic.form.submission.phone.invalid_country',
                ['%country%' => Countries::getName($countryCode)],
                'validators'
            );
```

Empty custom message suppresses default

When `country_validationmsg` is saved as an empty string, `??` treats it as present and passes an empty validation error to `failedValidation()`. Leaving the optional custom message blank should fall back to the translated default, so this should use an emptiness check or otherwise normalize empty custom messages before validation.

shinde-rahul reviewed (changes requested):
Rest looks good.
